boards/boardsource-blok/examples/blok_usb_serial.rs

```diff
@@ -0,0 +1,160 @@
+    hal::{
+        clocks::{init_clocks_and_plls, Clock},
+        pac,
+        pac::interrupt,
+        timer::Timer,
+        watchdog::Watchdog,
+        Sio,
+    },
+    Pins, XOSC_CRYSTAL_FREQ,
+};
+use core::fmt::Write;
+use heapless::String;
+use panic_halt as _;
+use usb_device::{
+    bus::UsbBusAllocator, device::UsbDevice, device::UsbDeviceBuilder, device::UsbVidPid,
+};
+use usbd_serial::SerialPort;
+
+// shared with the interrupt
+static mut USB_DEVICE: Option<UsbDevice<hal::usb::UsbBus>> = None;
+static mut USB_BUS: Option<UsbBusAllocator<hal::usb::UsbBus>> = None;
+static mut USB_SERIAL: Option<SerialPort<hal::usb::UsbBus>> = None;
+
+#[entry]
+fn main() -> ! {
+    let mut pac = pac::Peripherals::take().unwrap();
+
+    let mut watchdog = Watchdog::new(pac.WATCHDOG);
+
+    let clocks = init_clocks_and_plls(
+        XOSC_CRYSTAL_FREQ,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .ok()
+    .unwrap();
+
+    let sio = Sio::new(pac.SIO);
+    let _pins = Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    let _timer = Timer::new(pac.TIMER, &mut pac.RESETS);
+
+    let usb_bus = UsbBusAllocator::new(hal::usb::UsbBus::new(
+        pac.USBCTRL_REGS,
+        pac.USBCTRL_DPRAM,
+        clocks.usb_clock,
+        true,
+        &mut pac.RESETS,
+    ));
+    unsafe {
+        USB_BUS = Some(usb_bus);
+    }
+
+    let bus_ref = unsafe { USB_BUS.as_ref().unwrap() };
+
+    let serial = SerialPort::new(bus_ref);
+    unsafe {
+        USB_SERIAL = Some(serial);
+    }
+
+    let usb_device = UsbDeviceBuilder::new(bus_ref, UsbVidPid(0x1209, 0x0001))
+        .product("serial port")
+        .device_class(2) // from: https://www.usb.org/defined-class-codes
+        .build();
+    unsafe {
+        USB_DEVICE = Some(usb_device);
+    }
+
+    unsafe {
+        pac::NVIC::unmask(hal::pac::Interrupt::USBCTRL_IRQ);
+    }
+
+    let core = pac::CorePeripherals::take().unwrap();
+    let mut delay = cortex_m::delay::Delay::new(core.SYST, clocks.system_clock.freq().to_Hz());
+
+    let mut i: u8 = 0;
+
+    loop {
+        delay.delay_ms(2_000);
+
+        let mut text: String<20> = String::new();
+        writeln!(&mut text, "loop number: {}\r\n", i).unwrap();
+
+        write_serial(text.as_bytes());
+
+        i += 1;
+        if i >= 10 {
+            hal::rom_data::reset_to_usb_boot(0, 0);
+        }
+    }
+}
+
+/// Writes to the serial port.
+///
+/// We do this with interrupts disabled, to avoid a race hazard with the USB IRQ.
+fn write_serial(byte_array: &[u8]) {
+    let _ = critical_section::with(|_| unsafe {
+        USB_SERIAL.as_mut().map(|serial| serial.write(byte_array))
+    })
+    .unwrap();
+}
+
+/// This function is called whenever the USB Hardware generates
+/// an Interrupt Request
+#[allow(non_snake_case)]
+#[interrupt]
+unsafe fn USBCTRL_IRQ() {
```

Comment:
IIRC, this is not sufficient to remain compliant with USB's spec.
There also need to be a time based interrupt to garanty the IRQ is polled at least every 10ms.

See: https://docs.rs/usb-device/latest/usb_device/device/struct.UsbDevice.html#method.poll

Reply:
It says there that it should be called "preferably from an interrupt handler.". And I'm not sure where you read that it's not sufficient to remain compliant with USB's spec and what you mean by that. I'm quite new to programming, please help me out.

Reply:
It should be called preferably from an interrupt handler in order to minimise the latency between the event and its processing.
The next sentence is:

> Must be called at least once every 10 milliseconds while connected to the USB host to be USB compliant.

The USB interrupt is not guaranteed to be called at any interval (eg IIRC during usb-suspend, there's no USB interrupt).
The `usb-device`'s stack may need to update its state machine eg to handle timeouts, disconnection or some internal resource allocation.
It may even not currently (or no longer) strictly need it, but still keep this requirement to future-proof the crate 'just in case'.

Reply:
Sorry for all the questions and thank you for your time.

If I understand correctly the usb interrupt shouldn't be used to poll the `usb-device` because it has to keep getting updated even if the host doesn't request any data.
But shouldn't the [pico serial interrupt example](https://github.com/rp-rs/rp-hal-boards/blob/main/boards/rp-pico/examples/pico_usb_serial_interrupt.rs) then be changed as well?
Would this mean that usb interrupt shouldn't be used at all and the `usb-device` has to be called as often as possible in the main loop?
How would you then allow e.g. sleeping in the main loop?

I'm once again sorry, but i don't understand what "it" and "this" refer to in your last sentence.

Reply:
> If I understand correctly the usb interrupt shouldn't be used to poll the usb-device because it has to keep getting updated even if the host doesn't request any data.

The usb-interrupt **should** be used to poll the `usb-device` to guaranty minimal latency but it should also be polled on a time based schedule. It could be from an interrupt or from the main loop.

> But shouldn't the [pico serial interrupt example](https://github.com/rp-rs/rp-hal-boards/blob/main/boards/rp-pico/examples/pico_usb_serial_interrupt.rs) then be changed as well?

It's well possible. Some of the examples predate the `critical-section` support etc.

>  How would you then allow e.g. sleeping in the main loop?

If you want to sleep in the main loop but still wake at set interval, you can use the [`cortex_m::asm::wfe`](https://docs.rs/cortex-m/latest/cortex_m/asm/fn.wfe.html) or [`cortex_m::asm::wfi`](https://docs.rs/cortex-m/latest/cortex_m/asm/fn.wfi.html) and configure your system accordingly to generate an event (via timer, alarm or systick) an wake the core.

> I'm once again sorry, but i don't understand what "it" and "this" refer to in your last sentence.

My apologies, let me rephrase that sentence:

> The _**`usb-device` stack**_ may even not currently (or no longer) strictly need _**this time-based polling**_, but still keep this requirement _**(the time-based polling)**_ to future-proof the `usb-device` crate 'just in case'.

I hope this all help.